File: Binary Tree/BuildTree.cpp
#include <bits/stdc++.h>
using namespace std;

// A Node consists of: left pointer, value, and right pointer
class Node {
public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};

class Tree {
private:
    Node* root;

public:
    Tree() {
        root = NULL;
    }

    void BuildTree(string str) {
        if (str.empty() || str[0] == 'N') return;

        vector<string> info;
        stringstream ss(str);
        string word;
        while (ss >> word) {
            info.push_back(word);
        }

        int val = stoi(info[0]);
        root = new Node(val);

        queue<Node*> Q;
        Q.push(root);

        int i = 1;
        while (!Q.empty() && i < info.size()) {
            Node* curr = Q.front();
            Q.pop();

            // Left child
            if (i < info.size() && info[i] != "N") {
                val = stoi(info[i]);
                curr->left = new Node(val);
                Q.push(curr->left);
            }
            i++;

            // Right child
            if (i < info.size() && info[i] != "N") {
                val = stoi(info[i]);
                curr->right = new Node(val);
                Q.push(curr->right);
            }
            i++;
        }
    }

    // Utility function to print inorder traversal (for testing)
    void inorder(Node* node) {
        if (!node) return;
        inorder(node->left);
        cout << node->data << " ";
        inorder(node->right);
    }

    // Public method to trigger traversal
    void printInorder() {
        inorder(root);
        cout << endl;
    }
};

int main() {
    Tree t1;
    string str;
    getline(cin, str);  // e.g., "10 20 30 N 40 N N"

    t1.BuildTree(str);

    cout << "Inorder Traversal: ";
    t1.printInorder();

    return 0;
}

File: Binary Tree/Build_Tree.py
from collections import deque

# Define a Node class
class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None

# Tree class to build tree and perform inorder traversal
class Tree:
    def __init__(self):
        self.root = None

    def build_tree(self, s):
        if not s or s[0] == 'N':
            return

        info = s.split()
        self.root = Node(int(info[0]))

        queue = deque([self.root])
        i = 1

        while queue and i < len(info):
            curr = queue.popleft()

            # Left child
            if i < len(info) and info[i] != 'N':
                curr.left = Node(int(info[i]))
                queue.append(curr.left)
            i += 1

            # Right child
            if i < len(info) and info[i] != 'N':
                curr.right = Node(int(info[i]))
                queue.append(curr.right)
            i += 1

    def inorder(self, node):
        if not node:
            return
        self.inorder(node.left)
        print(node.data, end=" ")
        self.inorder(node.right)

    def print_inorder(self):
        self.inorder(self.root)
        print()

# Main code
if __name__ == "__main__":
    tree = Tree()
    s = input("Enter tree in level order (e.g., '10 20 30 N 40 N N'): ")
    tree.build_tree(s)

    print("Inorder Traversal:", end=" ")
    tree.print_inorder()
